src/org/jetbrains/java/decompiler/modules/decompiler/IfPatternMatchProcessor.java

```diff
@@ -149,4 +197,91 @@ private static void findVarsInPredecessors(List<VarVersionPair> vvs, Statement r
       }
     }
   }
+
+  /**
+   * Gets the last guaranteed executed exprent in an expression.
+   * @param ifTrue if true, gets the last executed exprent when the condition is true.
+   *               if false, gets the last executed exprent when the condition is false.
+   * @param onlyIfTrue if true, only returns the last executed exprent if the exprent had to return true for
+   *                  the requested outcome to be selected.
+   * @return the last executed exprent
+   */
+  public static Exprent getLastExprentWhen(Exprent base, boolean ifTrue, boolean onlyIfTrue) {
+    switch (base.type){
+      case FUNCTION: {
+        FunctionExprent func = (FunctionExprent) base;
+        switch (func.getFuncType()) {
+          case BOOLEAN_AND: {
+            if (ifTrue) {
+              // when `&&` returns true, the second exprent had to run and return true
+              return getLastExprentWhen(func.getLstOperands().get(1), true, onlyIfTrue);
+            }
+            // when `&&` returns false, either could have returned false, so we go to
+            // the default case of returning ourselves
+            break;
+          }
+          case BOOLEAN_OR: {
+            if (!ifTrue) {
+              // when `||` returns false, the second exprent had to run and return false
+              return getLastExprentWhen(func.getLstOperands().get(1), false, onlyIfTrue);
+            }
+            // when `||` returns true, either could have returned true, so we go to
+            // the default case of returning ourselves
+            break;
+          }
+          case BOOL_NOT: {
+            // when `!` returns true, the exprent had to return false
+            // when `!` returns false, the exprent had to return true
+            return getLastExprentWhen(func.getLstOperands().get(0), !ifTrue, onlyIfTrue);
+          }
+
+          // TEMPORARY
+          // (Nothing is as permanent as a temporary solution :p)
```

Comment:
Clean up comment here

Reply:
Yeah, I'm just not sure what to do with it

Reply:
I think you can remove the `(Nothing is as permanent as a temporary solution :p)` and the `:/` at the end there, that'd make it okay with me